<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Penetrance Estimation for Family-Based Cancer Penetrance Studies • penetrance</title>
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Penetrance Estimation for Family-Based Cancer Penetrance Studies">
<meta name="description" content="An R package for Penetrance Estimation.">
<meta property="og:description" content="An R package for Penetrance Estimation.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">penetrance</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/simulation_study_real.html">Simulation study from empirical data with penetrance</a></li>
    <li><a class="dropdown-item" href="articles/simulation_study.html">Simulation Study with penetrance</a></li>
    <li><a class="dropdown-item" href="articles/using_penetrance.html">Using penetrance</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/nicokubi/penetrance/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="penetrance-r-package">Penetrance R package<a class="anchor" aria-label="anchor" href="#penetrance-r-package"></a>
</h1></div>
<p>An R package for the estimation of age-specific penetrance for complex family-based studies in a format compatible with with the PanelPRO R package.</p>
<div class="section level2">
<h2 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h2>
<p>Accurate estimation of age-specific penetrance is essential for assessing cancer risk in individuals with pathogenic genetic variants (PGVs). Penetrance refers to the probability that an individual carrying a specific genetic variant will develop the associated trait, such as cancer. Estimating this probability is a crucial step in clinical decision-making and personalized risk assessment for hereditary (cancer) syndromes.</p>
<p>The package leverages Mendelian inheritance models, which are widely used in family-based genetic studies to assess how genetic variants are passed down through generations. These models typically involve a —an individual for whom family history and genetic data are collected. The proband serves as the starting point for mapping out the family’s genetic structure, including relationships and phenotypic traits, such as cancer diagnoses. Family data, including cancer occurrence, ages of diagnosis, and genetic test results, are collected for the proband and their relatives. Using this data, Mendelian models compute the likelihood of certain genetic configurations and disease outcomes based on inheritance patterns.</p>
<p>The core methodology in the package relies on a four-parameter Weibull distribution to model age-specific penetrance. Estimation is performed using a Bayesian framework with Markov Chain Monte Carlo (MCMC) methods, allowing the package to provide robust and flexible penetrance estimates. Through this approach, the package models the likelihood of cancer occurrence across family members, even when some genotypic information is missing or incomplete, which is common in real-world studies.</p>
<p>The package also incorporates prior knowledge into the estimation process, enabling users to specify default, custom, or study-based prior distributions. By employing the Elston-Stewart peeling algorithm, the package efficiently calculates likelihoods across family pedigrees, ensuring scalability and accuracy, even in large datasets.</p>
<p>By providing user-friendly functions for data input, prior specification, and estimation, the package equips researchers and clinicians with a powerful tool for estimating cancer risk in complex family-based studies. This empowers informed decision-making and preventive strategies in hereditary cancer syndromes, where understanding the genetic basis of risk is critical for patient care.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install, use</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>git clone git<span class="sc">@</span>github.com<span class="sc">:</span>nicokubi<span class="sc">/</span>penetrance.git</span></code></pre></div>
<p>Open the source directory as new R project and install the package with</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install</span>()</span></code></pre></div>
<p>or directly in R studio</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"https://github.com/nicokubi/penetrance"</span>)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start-guide">Quick-start guide<a class="anchor" aria-label="anchor" href="#quick-start-guide"></a>
</h2>
<p>This following is a quick-start guide for basic usage of the package. For greater detail on options, please refer to the other articles.</p>
<p>The primary function in the package is <code>penetrance</code>. The package workflow includes three main parts: user input, including family data in the form of pedigrees and specification for the penetrance estimation, the estimation of the posterior distribution using the MCMC approach, and the outputting of the results in the form of the samples from the approximated posterior distribution, i.e. the estimated penetrance function.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/nicokubi/penetrance" class="external-link">penetrance</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="pedigree">Pedigree<a class="anchor" aria-label="anchor" href="#pedigree"></a>
</h3>
<p>The user must specify the <code>pedigree</code> argument as a data frame that contains the family data (see <code>test_family_1</code>). The family data must be in the correct format with the following columns:</p>
<ul>
<li><p><code>ID</code>: A numeric value representing the unique identifier for each individual. There should be no duplicated entries.</p></li>
<li><p><code>Sex</code>: A numeric value where <code>0</code> indicates female and <code>1</code> indicates male. Missing entries are not currently supported.</p></li>
<li><p><code>MotherID</code>: A numeric value representing the unique identifier for an individual’s mother.</p></li>
<li><p><code>FatherID</code>: A numeric value representing the unique identifier for an individual’s father.</p></li>
<li><p><code>isProband</code>: A numeric value where <code>1</code> indicates the individual is a proband and <code>0</code> otherwise.</p></li>
<li><p><code>CurAge</code>: A numeric value indicating the age of censoring (current age if the person is alive or age at death if the person is deceased). Allowed ages range from <code>1</code> to <code>94</code>.</p></li>
<li><p><code>isAff</code>: A numeric value indicating the affection status of cancer, with <code>1</code> for diagnosed individuals and <code>0</code> otherwise. Missing entries are not supported.</p></li>
<li><p><code>Age</code>: A numeric value indicating the age of cancer diagnosis, encoded as <code>NA</code> if the individual was not diagnosed. Allowed ages range from <code>1</code> to <code>94</code>.</p></li>
<li><p><code>Geno</code>: A column for germline testing or tumor marker testing results. Positive results should be coded as <code>1</code>, negative results as <code>0</code>, and unknown results as <code>NA</code> or left empty.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="model-specification">Model specification<a class="anchor" aria-label="anchor" href="#model-specification"></a>
</h3>
<p>There are a few ways in which a user can specify how the estimation approach is run. Available options are:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' @param pedigree A data frame containing the pedigree data in the required format. It should include the following columns:</span></span>
<span><span class="co">#'   - `PedigreeID`:  A numeric value representing the unique identifier for each family There should be no duplicated entries.</span></span>
<span><span class="co">#'   - `ID`: A numeric value representing the unique identifier for each individual. There should be no duplicated entries.</span></span>
<span><span class="co">#'   - `Sex`: A numeric value where `0` indicates female and `1` indicates male. Missing entries are not currently supported.</span></span>
<span><span class="co">#'   - `MotherID`: A numeric value representing the unique identifier for an individual's mother.</span></span>
<span><span class="co">#'   - `FatherID`: A numeric value representing the unique identifier for an individual's father.</span></span>
<span><span class="co">#'   - `isProband`: A numeric value where `1` indicates the individual is a proband and `0` otherwise.</span></span>
<span><span class="co">#'   - `CurAge`: A numeric value indicating the age of censoring (current age if the person is alive or age at death if the person is deceased). Allowed ages range from `1` to `94`.</span></span>
<span><span class="co">#'   - `isAff`: A numeric value indicating the affection status of cancer, with `1` for diagnosed individuals and `0` otherwise. Missing entries are not supported.</span></span>
<span><span class="co">#'   - `Age`: A numeric value indicating the age of cancer diagnosis, encoded as `NA` if the individual was not diagnosed. Allowed ages range from `1` to `94`.</span></span>
<span><span class="co">#'   - `Geno`: A column for germline testing or tumor marker testing results. Positive results should be coded as `1`, negative results as `0`, and unknown results as `NA` or left empty.</span></span>
<span><span class="co">#' @param twins Identical twins or triplets in the family can be specifed. For example, to indicate that `ora024` and `ora027` are identical twins, and so are `aey063` and `aey064`, </span></span>
<span><span class="co">#' then we can use the following as the twins arguement: twins &lt;- list(c("ora024", "ora027"), c("aey063", "aey064"))</span></span>
<span><span class="co">#' @param n_chains Number of chains for parallel computation. Default is 1.</span></span>
<span><span class="co">#' @param n_iter_per_chain Number of iterations for each chain. Default is 10000.</span></span>
<span><span class="co">#' @param ncores Number of cores for parallel computation. Default is 6.</span></span>
<span><span class="co">#' @param baseline_data Data for the baseline risk estimates (probability of developing cancer), such as population-level risk from a cancer registry. Default data, for exemplary purposes, is for Colorectal cancer from the SEER database.</span></span>
<span><span class="co">#' @param max_age Maximum age considered for analysis. Default is 94.</span></span>
<span><span class="co">#' @param removeProband Logical, indicating whether to remove probands from the analysis. Default is FALSE.</span></span>
<span><span class="co">#' @param ageImputation Logical, indicating whether to perform age imputation. Default is FALSE.</span></span>
<span><span class="co">#' @param median_max Boolean indicating whether to use the baseline median age or max_age as an upper bound for the median proposal. Default is TRUE.</span></span>
<span><span class="co">#' @param BaselineNC Boolean indicating that the non-carrier penetrance is assumed to be the baseline penetrance. Default is TRUE.</span></span>
<span><span class="co">#' @param var Vector of variances for the proposal distribution. Default is c(0.1, 0.1, 2, 2, 5, 5, 5, 5).</span></span>
<span><span class="co">#' @param burn_in Fraction of results to discard as burn-in (0 to 1). Default is 0 (no burn-in).</span></span>
<span><span class="co">#' @param thinning_factor Factor by which to thin the results. Default is 1 (no thinning).</span></span>
<span><span class="co">#' @param distribution_data Data for generating prior distributions.</span></span>
<span><span class="co">#' @param af Allele frequency for the risk allele. Default is 0.0001.</span></span>
<span><span class="co">#' @param max_penetrance Maximum penetrance considered for analysis. Default is 1.</span></span>
<span><span class="co">#' @param sample_size Optional sample size for distribution generation.</span></span>
<span><span class="co">#' @param ratio Optional ratio parameter for distribution generation.</span></span>
<span><span class="co">#' @param prior_params Parameters for prior distributions.</span></span>
<span><span class="co">#' @param risk_proportion Proportion of risk for distribution generation.</span></span>
<span><span class="co">#' @param summary_stats Boolean indicating whether to include summary statistics in the output. Default is TRUE.</span></span>
<span><span class="co">#' @param rejection_rates Boolean indicating whether to include rejection rates in the output. Default is TRUE.</span></span>
<span><span class="co">#' @param density_plots Boolean indicating whether to include density plots in the output. Default is TRUE.</span></span>
<span><span class="co">#' @param plot_trace Boolean indicating whether to include trace plots in the output. Default is TRUE.</span></span>
<span><span class="co">#' @param penetrance_plot Boolean indicating whether to include penetrance plots in the output. Default is TRUE.</span></span>
<span><span class="co">#' @param penetrance_plot_pdf Boolean indicating whether to include PDF plots in the output. Default is TRUE.</span></span>
<span><span class="co">#' @param probCI Probability level for confidence intervals in penetrance plots. Default is 0.95.</span></span>
<span><span class="co">#' @param sex_specific Logical, indicating whether to use sex-specific parameters in the analysis. Default is TRUE.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="prior-specification">Prior Specification<a class="anchor" aria-label="anchor" href="#prior-specification"></a>
</h3>
<p>Penetrance provides the user with a flexible approach to prior specification, balancing customization with an easy-to-use workflow. In addition to providing default prior distributions, the package allows users to customize the priors by including existing penetrance estimates or prior knowledge. The following settings for the prior distribution specification are available: ### Additional User Inputs</p>
<ul>
<li><p>The <code>PenEstim</code> function takes baseline age-specific probabilitie of developing cancer as as input <code>baseline_data</code>. In the default setting with <code>BaselineNC = TRUE</code> this baseline is assumed to reflect the non-carrier penetrance. For rare mutations this is considered a reasonable assumption. The baseline_data must be a data frame with baseline data for females and males.</p></li>
<li><p>Furthermore the specification of the allele frequency of the pathogenic variant (<code>af</code>) is required.</p></li>
<li><p>The PenEsim also includes an option for automatic age imputation <code>AgeImputation</code>. We apply an age imputation as part of the MCMC routine. The imputation of ages is performed based on the individual’s affected status (<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>a</mi><mi>f</mi><mi>f</mi></mrow><annotation encoding="application/x-tex">aff</annotation></semantics></math>), sex (<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>s</mi><mi>e</mi><mi>x</mi></mrow><annotation encoding="application/x-tex">sex</annotation></semantics></math>), and their degree of relationship to the proband who is a carrier of the PV. For greater detail on the age imputation approach see documentation.</p></li>
<li><p>For the likelihood calculation monozygous twins can be specified using the <code>twins</code> arguement.</p></li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">twins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ora024"</span>, <span class="st">"ora027"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"aey063"</span>, <span class="st">"aey064"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/nicokubi/penetrance/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing penetrance</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>BayesMendel Lab <br><small class="roles"> Author, maintainer </small>  </li>
<li>Nicolas Kubista <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by BayesMendel Lab, Nicolas Kubista.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
